File: top_interview_questions/python/230_kth_smaillest_element_in_a_bst.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: Fizz Cao 
@file: 230_kth_smaillest_element_in_a_bst.py
@time: 26/09/2021 10:53 pm
'''
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        ans, stack = 0, []

        while ((stack or root) and k):
            while root:
                stack.append(root)
                root = root.left
            ans = stack[-1].val
            k -= 1
            root = stack.pop().right

        return ans

File: top_interview_questions/java/230_kth_smaillest_element_in_a_bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int ans = 0;
        LinkedList<TreeNode> stack = new LinkedList ();

        while ((root != null || stack.size() > 0) && k > 0) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            ans = stack.peek().val;
            k -= 1;
            root = stack.pop().right;
        }
        return ans;
    }
}